service/pager/query_pagers.cc

```diff
@@ -272,6 +274,38 @@ class filtering_query_pager : public query_pager {
     }
 };
 
+class ghost_row_deleting_query_pager : public service::pager::query_pager {
+    service::storage_proxy& _proxy;
+    db::timeout_clock::duration _timeout_duration;
+public:
+    ghost_row_deleting_query_pager(schema_ptr s, shared_ptr<const cql3::selection::selection> selection,
+                service::query_state& state,
+                const cql3::query_options& options,
+                lw_shared_ptr<query::read_command> cmd,
+                dht::partition_range_vector ranges,
+                cql3::cql_stats& stats,
+                service::storage_proxy& proxy,
+                db::timeout_clock::duration timeout_duration)
+        : query_pager(proxy, s, selection, state, options, std::move(cmd), std::move(ranges))
+        , _proxy(proxy)
+        , _timeout_duration(timeout_duration)
+    {}
+    virtual ~ghost_row_deleting_query_pager() {}
+
+    virtual future<result<>> fetch_page_result(cql3::selection::result_set_builder& builder, uint32_t page_size, gc_clock::time_point now, db::timeout_clock::time_point timeout) override {
+        return do_fetch_page(page_size, now, timeout).then(utils::result_wrap([this, &builder, page_size, now] (service::storage_proxy::coordinator_query_result qr) {
+            _last_replicas = std::move(qr.last_replicas);
+            _query_read_repair_decision = qr.read_repair_decision;
+            qr.query_result->ensure_counts();
+            return seastar::async([this, query_result = std::move(qr.query_result), page_size, now] () mutable -> result<> {
+                handle_result(db::view::delete_ghost_rows_visitor{_proxy, _state, view_ptr(_schema), _timeout_duration},
+                        std::move(query_result), page_size, now);
+                return bo::success();
+            });
+        }));
+    }
```

Comment:
Ideally, what should be exposed for callers is just passing the visitor, or a lambda which wraps a visitor. It will require quite a bit of refactoring though, so I decided it can be stalled until we need to add yet another such pager. Which might be very soon, if we want to proceed with the other part of repairing views - regenerating missing view information from given base ranges.

Reply:
sure.

Reply:
I attempted a patch today, but needed to stop before the result got way too ugly. I tried abstracting out a "page handler" responsible for handling a single page of results. With such a handler, our `query_pager` class stops being virtual, and all the implementation is moved to the handler. Thing is, that creates another layer of indirection - we currently usually operate on unique_ptr<pager> in order to make liveness issues easier (pager callsites can't be easily coroutinized due to miscompilations, it has been tried multiple times). Storing yet another pointer-to-implementation is simply one more allocation. But what's worse, current page handling code actually uses most of query_pager's internal state, so we would either need to allow getting this state, or create an ugly list of friend classes (and since friendliness is not inherited, each subclass would have to be explicitly named, which is not really the purpose of pimpl). So, to sum up, current interface of inheriting from query_pager and overriding a single virtual method is not that bad... except that it indeed sounds weird to have a "ghost row deleting query pager".